### PR TITLE
[VM][Tools] Run Generated Programs on VM Runtime

### PR DESCRIPTION
**Pull request by darioncassel**
## Motivation

This commit adds the ability to run generated bytecode programs on the Move VM runtime.
Additionally, it makes explicit the functions that depend on memory safety checks so
that they can be disabled if generation of only valid programs is desired.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Local testing with `cargo test` and safety analysis with MIRAI.

## Related PRs

N/A

### CODE REVIEW

File: language/tools/test_generation/src/abstract_state.rs [hunk 1]
```diff
@@ -99,6 +99,9 @@ pub struct AbstractState {
 
     /// The module state
     pub module: CompiledModule,
+
+    /// This flag is set when applying an instruction should result in an error
```
Comment (hermanventer) [inline language/tools/test_generation/src/abstract_state.rs]: I don't understand this comment. Please try to clarify it.

File: language/tools/test_generation/src/bytecode_generator.rs [hunk 2]
```diff
@@ -343,10 +349,11 @@ impl BytecodeGenerator {
         return_len: usize,
         state: &AbstractState,
         candidates: &[(StackEffect, Bytecode)],
-    ) -> Bytecode {
+    ) -> Result<Bytecode, String> {
+        debug!("Candidates: {:?}", candidates);
```
Comment (hermanventer) [inline language/tools/test_generation/src/bytecode_generator.rs]: Did you mean to leave this here?

File: language/tools/test_generation/src/bytecode_generator.rs [hunk 3]
```diff
@@ -363,10 +370,10 @@ impl BytecodeGenerator {
             // Add candidates should not be empty unless the list of bytecode instructions is
             // changed
             if add_candidates.is_empty() {
-                panic!("Could not find valid candidate");
+                return Err("Could not find valid add candidate".to_string());
             }
             next_instruction_index = self.rng.gen_range(0, add_candidates.len());
-            add_candidates[next_instruction_index].1.clone()
+            Ok(add_candidates[next_instruction_index].1.clone())
```
Comment (hermanventer) [inline language/tools/test_generation/src/bytecode_generator.rs]: At this point I find myself wondering why you don't map out the stack effects from add_candidates. I also find it strange that you need to call clone again here.

File: language/tools/test_generation/src/bytecode_generator.rs [hunk 4]
```diff
@@ -378,25 +385,35 @@ impl BytecodeGenerator {
             // Sub candidates should not be empty unless the list of bytecode instructions is
             // changed
             if sub_candidates.is_empty() {
-                panic!("Could not find valid candidate");
+                return Err("Could not find sub valid candidate".to_string());
             }
             next_instruction_index = self.rng.gen_range(0, sub_candidates.len());
-            sub_candidates[next_instruction_index].1.clone()
+            Ok(sub_candidates[next_instruction_index].1.clone())
```
Comment (hermanventer) [inline language/tools/test_generation/src/bytecode_generator.rs]: Same here.

File: language/tools/test_generation/src/bytecode_generator.rs [hunk 5]
```diff
@@ -487,7 +512,10 @@ impl BytecodeGenerator {
         signature: &FunctionSignature,
         module: CompiledModuleMut,
     ) -> Vec<Bytecode> {
-        let number_of_blocks = 3;
+        let number_of_blocks = self.rng.gen_range(1, MAX_CFG_BLOCKS + 1);
+        // The number of basic blocks must be at least one based on the
+        // generation range.
+        assume!(number_of_blocks > 0);
```
Comment (hermanventer) [inline language/tools/test_generation/src/bytecode_generator.rs]: At least make an issue for MIRAI to add a prebuilt summary for gen_range.

File: language/tools/test_generation/src/abstract_state.rs (hunk 1 above)
Comment (darioncassel) [reply language/tools/test_generation/src/abstract_state.rs]: Fix in upcoming PR.

File: language/tools/test_generation/src/bytecode_generator.rs (hunk 2 above)
Comment (darioncassel) [reply language/tools/test_generation/src/bytecode_generator.rs]: Yes, that's left there as debug output.

File: language/tools/test_generation/src/bytecode_generator.rs (hunk 3 above)
Comment (darioncassel) [reply language/tools/test_generation/src/bytecode_generator.rs]: Good point, I've changed it to map those out. Unfortunately that clone can't be avoided.

File: language/tools/test_generation/src/bytecode_generator.rs (hunk 4 above)
Comment (darioncassel) [reply language/tools/test_generation/src/bytecode_generator.rs]: Same as above.

File: language/tools/test_generation/src/bytecode_generator.rs (hunk 5 above)
Comment (darioncassel) [reply language/tools/test_generation/src/bytecode_generator.rs]: Working on a PR for MIRAI.

**Pull request closed without merging**